src/AI-Algorithms-Graph/AIDinic.class.st

```diff
@@ -0,0 +1,232 @@
+	The level graph assigns levels or distances to each node, indicating the shortest path from the source.
+	The ret boolean indicates if there exists and augmenting path(path from source to sink) in the residual level graph."
+
+	| node ind ret |
+	[ queue isNotEmpty ] whileTrue: [
+		node := queue removeFirst.
+		ind := mapping at: node.
+		ind do: [ :i |
+			| e n |
+			e := edges at: i.
+			n := e to.
+			e capacity - e flow >= 1 & (n level = -1) ifTrue: [
+				n level: node level + 1.
+				queue addLast: n ] ] ].
+	ret := sink level == -1.
+	^ ret
+]
+
+{ #category : #configuration }
+AIDinic >> currentIndexSetup [
+
+	| nod |
+	nod := self nodes.
+	nod do: [ :n | n currentIndex: 1 ]
+]
+
+{ #category : #backtracking }
+AIDinic >> dfs: fromNode pushed: p [
+	"This method uses DFS-style algorithm to find a blocking flow (A blocking flow is a flow that saturates all the edges on the path, preventing any further flow) along the augmenting path.
+This performs a series of depth-first searches on the residual graph, exploring the edges with positive residual capacity."
+
+	| arr cid id edg reverseEdg toNode tr min |
+	p == 0 ifTrue: [ ^ 0 ].
+	fromNode == sink ifTrue: [ ^ p ].
+
+	arr := mapping at: fromNode.
+	cid := fromNode currentIndex.
+	tr := 0.
+
+	[ cid <= arr size ] whileTrue: [
+		id := arr at: cid.
+		edg := edges at: id.
+		toNode := edg to.
+		"Checking if the TO NODE is one level greater than FROM NODE, and for positive residual capacity"
+		fromNode level + 1 == toNode level & (edg capacity - edg flow >= 1)
+			ifTrue: [
+				fromNode currentIndex: cid.
+				"Calculating the minimum value between the current positive residual capacity and the pushed value from previous iterations"
+				min := self minimumValue: edg capacity - edg flow compare: p.
+				"The recursive call of DFS"
+				tr := self dfs: toNode pushed: min ].
+
+		tr == 0 ifFalse: [
+			edg flow: edg flow + tr.
+			reverseEdg := self reverseEdge: toNode to: fromNode.
+			reverseEdg flow: reverseEdg flow - tr.
+			^ tr ].
+		cid := cid + 1.
+		fromNode currentIndex: cid ].
+	^ 0
+]
+
+{ #category : #configuration }
+AIDinic >> edgeClass [
+
+	^ AINetworkFLowEdge
+]
+
+{ #category : #initialization }
+AIDinic >> initialize [ 
+	super initialize.
+	mapping:= Dictionary new.
+	queue := LinkedList new.
+]
+
+{ #category : #configuration }
+AIDinic >> initializeAndBfs [
+	"This method 
+ - Initialises all the levels and current index of the nodes
+ - Runs the bfs method to get the level graph
+ - Returns a boolean based on whether the there exists an augmenting path or not."
+
+	| bfsBool |
+	self levelSetup.
+	self currentIndexSetup.
+	start level: 0.
+	queue addLast: start.
+	bfsBool := self bfs.
+	^ bfsBool
+]
+
+{ #category : #configuration }
+AIDinic >> levelSetup [
+
+	| nod |
+	nod := self nodes.
+	nod do: [ :n | n level: -1 ]
+]
+
+{ #category : #configuration }
+AIDinic >> mapConfig [
+
+	| edg nod arr fromNode |
+	edg := self edges.
```

Comment:
I think that edg and nod are not needed. 
Now if they are please renamed them as edges and nodes to indicate that they are collections.